src/file/maps.h

```diff
@@ -0,0 +1,155 @@
+    .type = BPF_MAP_TYPE_HASH,
+    .key_size = sizeof(u64),
+    .value_size = sizeof(incomplete_file_message_t),
+    .max_entries = 2048,
+    .pinning = 0,
+    .namespace = "",
+};
+
+static __always_inline void enter_file_message(void *ctx, file_message_type_t kind)
+{
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    incomplete_file_message_t event = {0};
+    event.start_ktime_ns = bpf_ktime_get_ns();
+    event.kind = kind;
+
+    int ret = bpf_map_update_elem(&incomplete_file_messages, &pid_tgid, &event, BPF_ANY);
+    if (ret < 0)
+    {
+        file_message_t fm = {0};
+        fm.type = FM_WARNING;
+        fm.u.warning.pid_tgid = pid_tgid;
+        fm.u.warning.message_type.file = kind;
+        fm.u.warning.code = W_UPDATE_MAP_ERROR;
+        fm.u.warning.info.err = ret;
+
+        push_file_message(ctx, &fm);
+    }
+}
+
+static __always_inline incomplete_file_message_t* get_event(file_message_type_t kind, u64 *pid_tgid) {
+  incomplete_file_message_t *event = bpf_map_lookup_elem(&incomplete_file_messages, pid_tgid);
+  if (event == NULL) return NULL;
+
+  if (event->kind != kind) {
+    // TODO: emit warning
+    bpf_map_delete_elem(&incomplete_file_messages, pid_tgid);
+    return NULL;
+  }
+
+  return event;
+}
+
+static __always_inline incomplete_file_message_t* get_current_event(file_message_type_t kind) {
+  u64 pid_tgid = bpf_get_current_pid_tgid();
+  return get_event(kind, &pid_tgid);
+}
+
+static __always_inline incomplete_file_message_t* set_file_path(struct pt_regs *ctx, file_message_type_t kind,
+                                                         void *path, void *dentry)
+{
+    incomplete_file_message_t* event = get_current_event(kind);
+    if (event == NULL) return NULL;
+    if (event->target_dentry != NULL) return NULL;
+
+    event->target_dentry = dentry;
+    event->vfsmount = read_field_ptr(path, CRC_PATH_MNT);
+    if (event->vfsmount == NULL) goto EmitWarning;
+
+    return event;
+
+ EmitWarning:;
+    file_message_t fm = {0};
+    push_file_warning(ctx, &fm, event->kind);
+    return NULL;
+}
+
+// Handles the end states of a given message
+// If null, do nothing
+// If warning, emit the warning
+// If success, tail call.
+static __always_inline void finish_message(void *ctx, file_message_t *fm)
+{
+    if (fm == NULL) return;
+    if (fm->type == FM_WARNING) goto EmitWarning;
+    bpf_tail_call(ctx, &tp_programs, FILE_PATHS);
+
+    // save the message type
+    fm->u.warning.message_type.file = fm->type;
+    error_info_t info = {0};
+    info.tailcall = FILE_PATHS;
+    set_local_warning(W_TAIL_CALL_MAX, info);
+EmitWarning:;
+    push_file_warning(ctx, fm, fm->u.warning.message_type.file);
+}
+
+#define GET_EXIT_EVENT(kind)                                            \
+    u64 pid_tgid = bpf_get_current_pid_tgid();                          \
+    incomplete_file_message_t *event = get_event(kind, &pid_tgid);      \
```

Comment:
Because the callers of `GET_EXIT_EVENT` (`POP_AND_SETUP` and `POP_AND_SETUP_ARGS`) need to use the `pid_tgid` later. If we called for `get_current_event` then we wouldn't have access to that `pid_tgid` and we would have to re-ask for it. Not a big deal, but might as well save instructions when it's easy.